### PR TITLE
Refactor saved reports

### PR DESCRIPTION
**Pull request by nickpell**
At some point in the process of learning how the post method for saved reports works I decided it would be easier to just refactor it myself.  Actually served as a useful learning exercise and probably faster than just reading the code.

Might make sense to merge after UCR saved reports so they can be tested one after the other.

@czue

### CODE REVIEW

File: corehq/apps/reports/forms.py [hunk 1]
```diff
@@ -3,7 +3,77 @@
...
+        widget=forms.HiddenInput(),
+    )
+    days = forms.IntegerField(
+        required=False,
+        widget=forms.HiddenInput(),
+    )
+    _id = forms.CharField(
+        required=False,
+        widget=forms.HiddenInput(),
+    )
```
Comment (millerdev) [inline corehq/apps/reports/forms.py]: Is this `_id` field possibly used to look up and update an existing record? If so, is there a permissions check to verify that the user is updating a record that they own and not someone elses?

File: corehq/apps/reports/views.py [hunk 2]
```diff
@@ -479,65 +484,80 @@ def touch_saved_reports_views(user, domain):
...
+
+    @property
+    @memoized
+    def config(self):
+        config = ReportConfig.get_or_create(
+            self.saved_report_config_form.cleaned_data['_id']
+        )
+        if config.owner_id:
+            # in case a user maliciously tries to edit another user's config
+            assert config.owner_id == self.user_id
```
Comment (millerdev) [inline corehq/apps/reports/views.py]: ah, I see, here's the check for that. Looks good.

File: corehq/apps/reports/views.py [hunk 3]
```diff
@@ -479,65 +484,80 @@ def touch_saved_reports_views(user, domain):
...
+            self.user_id,
+            self.post_data
+        )
 
-    for field in config.properties().keys():
-        if field in POST:
-            setattr(config, field, POST[field])
+    @property
+    def filters(self):
+        filters = deepcopy(self.post_data.get('filters', {}))
```
Comment (millerdev) [inline corehq/apps/reports/views.py]: Looks like a shallow copy would work here as well. They tend to be much cheaper.

**Comment by millerdev** [comment]
Did a pass, and looks good with possible exception of `deepcopy`. I'm signed off, but @czue should probably take a look before merging.

**Comment by czue** [comment]
@millerdev happy for you to merge if you're happy - this isn't really my area of expertise.

@nickpell couldn't tell whether your merge order comment was just a suggestion or you want to explicitly do it that way for QA reasons

**Comment by nickpell** [comment]
@millerdev fixed ```deepcopy```

@czue I take it back - I was thinking about QA, but I'm now confident in my own testing

**Comment by nickpell** [comment]
@millerdev Actually I'd like to get this merged soon if possible - there's some existing bugs that this change would help me fix

**Pull request merged**